pkg/workspace/assert.go

```diff
@@ -0,0 +1,167 @@
+		if syntaxErr != nil {
+			return syntaxErr
+		}
+		for _, v := range validations {
+			err := v.Validate(node)
+			if err != nil {
+				a.violations = append(a.violations, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func newValidationsFromAssertValidateAnnotation(annotation template.NodeAnnotation, n yamlmeta.Node, thread *starlark.Thread) ([]Validation, error) {
+	var validations []Validation
+	validationPosition := createAssertAnnotationPosition(annotation.Position, n)
+
+	if len(annotation.Kwargs) != 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found keyword argument (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	if len(annotation.Args) == 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found no arguments (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	for _, arg := range annotation.Args {
+		validationTuple, ok := arg.(starlark.Tuple)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found: %v (by %v)", AnnotationAssertValidate, arg.String(), validationPosition.AsCompactString())
+		}
+		if len(validationTuple) != 2 {
+			return nil, fmt.Errorf("Expected @%v 2-tuple, but found tuple with length %v (by %v)", AnnotationAssertValidate, len(validationTuple), validationPosition.AsCompactString())
+		}
+		message, ok := validationTuple[0].(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have descriptive string as the first item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[0].Type(), validationPosition.AsCompactString())
+		}
+		lambda, ok := validationTuple[1].(starlark.Callable)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have a validation function as the second item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[1].Type(), validationPosition.AsCompactString())
+		}
+		validations = append(validations, Validation{
+			thread:   thread,
+			msg:      message.String(),
+			f:        lambda,
+			position: validationPosition,
+		})
+	}
+
+	return validations, nil
+}
+
+// Validation is an assertion from an @assert/validate annotation.
+// One @assert/validate annotation can have multiple Validations.
+type Validation struct {
+	thread   *starlark.Thread
+	msg      string
+	f        starlark.Callable
+	position *filepos.Position
+}
+
+// Validate runs the function from the Validation with the node's value as arguments.
+//
+// Returns an error if the function returns an error, or if the resulting value from the function is falsy.
+// Otherwise, returns nil.
+func (v Validation) Validate(node yamlmeta.Node) error {
+	var nodeValue starlark.Value
+	switch values := node.GetValues(); len(values) {
+	case 0: // run validation with no arguments
```

Comment:
`node.GetValues()` couldn't return an empty list of array or map items whose length could be zero because that would mean the map or array was annotated with `@assert/validate`, which you can not do. I think a change could be made so that `@assert/validate` errors/panics if we find it on a map or array. With a guard in place, I think it would be right to panic for cases other than `len(values)==1`.

Reply:
> With a guard in place, I think it would be right to panic for cases other than len(values)==1.

My intuition says this is a good starting place.